### PR TITLE
Kw/feature branch frontend changes

### PR DESCRIPTION
**Pull request by kellyawang**
### Description
Feature branch containing changes to the frontend listed in this issue:
https://github.com/IQSS/dataverse/issues/5603#issuecomment-572759009

This branch contains the following changes: 
* adds a new d3 library package 'd3-tip' which provides some convenient methods for styling tooltips
* sorts circles based on the `diff` field rather than placing them randomly
* formats tooltips differently depending on if a node is a dataset or a dataverse
* moves tooltip text logic into a helper method
* widens the svg slightly for ease of debugging
* adds "link" attribute to some of dataverses/datasets for testing purposes
* collapses nested dataset/dataverse tooltip descriptions if there are more than 5 nested
* refines "click" and "mouse" events
* [Responsiveness] On click outside the vis, hide any tooltips that are selected
* [Responsiveness] Re-size svg if window resized
* [Slider] Circle filtering hooked up to slider
* add some processing to display a tooltip to the left if the circle is close to the right edge of the window

### Screenshots
A dataverse with nested dataverses
*Note: I'm showing the diff field at the bottom of the tooltip for now, can hide this later*
image

A dataverse with nested datasets
image

Dataset
image

If there are more than 5 datasets in a dataverse, show a link to “all datasets” that goes to the Dataverse page: the same link destination as the title of the Dataverse the datasets are in.

Screen Shot 2020-03-11 at 1 23 42 AM

### CODE REVIEW

File: js/circlepack.js [hunk 1]
```diff
@@ -81,36 +98,29 @@ Circlepack.prototype.wrangleData = function(){
...
       //return d.size;
       return 10;
     })
     .sort(function(a, b) {
-      //console.log("b: " + b.value + " a:" + a.value);
-      // changes the orientation
-      //return b.value - a.value;
+      b_diff = b.data.diff === undefined ? 0 : b.data.diff
+      a_diff = a.data.diff === undefined ? 0 : a.data.diff
+      return a_diff - b_diff;
```
Comment (kellyawang) [inline js/circlepack.js:110]: @scolapasta 
To change the ordering of the circles, this sort function will need to be modified. 

The function(a, b) passed into `.sort()` simply sorts an array of circle nodes (ie. dataverses and datasets), comparing the `diff` value of each circle. To position the circles in a different order, say by `date`, you probably want to write a custom compare function that compares dates, instead of just integers.

For more information, see: 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
https://stackoverflow.com/questions/6567941/how-does-sort-function-work-in-javascript-along-with-compare-function

File: js/circlepack.js [hunk 2]
```diff
@@ -143,29 +155,235 @@ Circlepack.prototype.updateVis = function(){
...
-        .style("opacity", 0);
-    });
+      vis.hideTooltipOnMouseout(d, this)
+    })
+    .on("click", function(d) {
+      vis.showTooltipOnSelect(d, this);
+      // block circle clicks from propagating to the global click event handler at the top of updateVis
+      d3.event.stopPropagation()
+    })
+    .merge(vis.circles) // update vis with new data
```
Comment (kellyawang) [inline js/circlepack.js:176]: After the initial load of data for the visualization which happens in the `loadData()` function in `main.js`, if we filter or modify the data we want to show in the visualization, this function tells d3 how to re-render the vis. This is referred to as the "enter-update-exit pattern" in d3, if you're curious to read more.

eg. On the first load of the vis, the `.enter` and `.append` puts all the circle svgs on the page. When the slider changes, we need to filter out the data and redraw the circles accordingly. Without reloading the entire page, d3 does this by setting new attributes on any existing svg elements, and removing svg elements that are no longer represented in the data

Basically, enter = new elements to draw on the screen
update = existing elements on the screen to be modified when the data changes. 
exit = svg elements that should be deleted from the DOM based on the new data

FYI, the merge() function just combines the enter and update "selections" to reduce the amount of code you need to write. ([read here](https://stackoverflow.com/questions/47066905/d3-merge-function))

Comment (kellyawang) [reply js/circlepack.js:176]: I'm not sure if this would be useful, but you can find a basic intro in the book _Interactive Data Visualization for the Web: 2nd Edition_ by Scott Murray. There's a chapter on updating data and visuals with animated transitions that we learned about in Hanspeter's class (Chapter 9: Updates, Transitions, and Motion)

File: js/circlepack.js [hunk 3]
```diff
@@ -143,29 +155,235 @@ Circlepack.prototype.updateVis = function(){
...
+  if (isDataset(d)) {
+    title_html = `<div class="tooltip-dataset tooltip-title"><a href="${title_link}" target="_blank">${d.data.name}</a></div>`
+
+  // else if it's a dataverse, construct a dataverse title and add children to the description section
+  } else if (isDataverse(d)) {
+    title_html = `<div class="tooltip-dataverse tooltip-title"><a href="${title_link}" target="_blank">${d.data.name}</a></div>`
+    children_html = children_html.concat(`<ul>`)
+
+    // if dataverse has more than 5 children show link to all datasets
+    if (d.children.length > 5) {
```
Comment (kellyawang) [inline js/circlepack.js:369]: Update the logic in this if block here to loop over the first 5 nested datasets/dataverses before you add on the link to "All datasets"

**Comment by mheppler** [comment]
Based on conversations from 5/4 with @kellyawang, @TaniaSchlatter and team, I am going to merge this PR into the main repo, then submit another PR with UI clean up changes that I started on.

**Review by mheppler — APPROVED** [review]
Looks good. 👍
